### PR TITLE
Making k and num_candidates optional for knn search

### LINKED ISSUE #97533: Make k and num_candidates optional in the knn section
### Description

`k` and `num_candidates` are required parameters of the knn section. Could they have sensible defaults instead? E.g. could we default `k` to `size` and `num_candidates` to a value that would yield a sensible latency/recall trade-off depending on `k`?

### PR DESCRIPTION
**Pull request by pmpailis**
The aim of this PR is to make `num_candidates` and `k` optional for `knn` search, and provide instead some sensible defaults. 

As also defined [in the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search-api.html), `k` defines how many results we will eventually report back from each shard, while `num_candidates` is an exploration parameter and defines the size of the  nearest neighbors queue we will compute on each shard.  

`k` is used in two places: 
* On the collection phase of each shard (`DfsPhase#executeKnnVectorQuery`) when creating the `TopScoreDocCollector` for executing the knn query
* On the coordinator node, when merging the knn results from different shards (through the `SearchPhaseController#mergeKnnResults`) in order to keep the top `k`. 

`num_cands` is used solely within the `KnnSearchBuilder` when creating the appropriate `KnnVectorQueryBuilder` object. 

The parameter `k` can default to `size` (or `DEFAULT_SIZE ` if not present) - while for `num_cands` things are a bit more interesting :)  There are mainly 2 approaches that we could follow:
* Define it based on `k` (or `size`), finding a sweet-spot between recall / latency, ensuring that we will try to perform as little operations as possible 
* Define it based on the graph size, i.e. for smaller graphs have fewer `numCands` (always`>= k` though) - while for bigger ones we might want to allow for a few more steps to avoid local minima and try to find the true nearest neighbors. This could potentially give us more freedom in controlling (and reducing) the number of nodes we visit for each shard.

In this PR, as a first step, we will look into the 1st approach though, and make `numCands` depend on the incoming search request, instead of the indexed data. Based on the discussion & benchmarking results in the related issue, we'll proceed to define it as `num_candidates=Math.min(1.5 * k, 10_000)` (where 10_000 is the limit for `num_candidates`)

As there are some restrictions on the relationship between `numCands` and `k`, if we are to use any default values, in this PR we change the `k` and `numCands` variables from `final` and apply the same validation checks that are defined in the constructor to make sure that we either have a properly defined `KnnSearchBuilder` instance, or fail with an `IllegalArgumentException`. This is to also cover the scenario where a user might have provided `numCands` and `size` values, where `numCands < size`, so we want to raise an exception early on. 

In the future, should we decide to make `numCands` defined based on the graph size, we can move the implementation & assignment over to `KnnVectorQueryBuilder#doToQuery` where we have info on the specific shard/graph size. 

Closes https://github.com/elastic/elasticsearch/issues/97533

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java:453]: This is still TBD

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:169]: I wonder if we could handle this within the SearchSourceBuilder directly? When parsing (thus creating via an API), parsing the `knn` section or `size` section could set the `k` for the `knn` there.

Similarly if somebody calls `SearchSourceBuilder#setSize(int)` that could propagate down to the `knn` object if it exists?

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java:220]: when you are ready for further review, etc. all these will have to be backwards compatible. Meaning looking at the `in.getTransportVersion()` and verifying its at least the version where this change is introduced.

Similar logic down in `writeTo(StreamOutput)` method.

Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java:220]: Thanks for the heads up @benwtrent ! Will update accordingly.

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:169]: I was looking into it initially, but I wasn't impressed with a couple of approaches I've tried, as we were introducing dependencies between different component in the `SearchSourceBuilder` - especially when there's no guaranty of order.  On the other hand, as things are now, if someone calls `SearchSourceBuilder#setSize` at a later point, then they would also have the responsibility of setting up & updating `k` which is not right 🤔  (similarly for `org.elasticsearch.search.DefaultSearchContext#size(int) `) 

Will look into it a bit more.

**Comment by pmpailis** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/TransportVersions.java
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/TransportVersions.java:196]: Will update this to the proper version just before merging the PR

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: Updated the functionality to go through the `SearchSourceBuilder` in https://github.com/elastic/elasticsearch/pull/101209/commits/9092f378c9c8ad8d2b5aebc661bf2232fd712160. Haven't set it in the `SearchSourceBuilder#size` method, but rather when we parse the initial `size` param (iff present).  

Happy to update it as well, but was wondering which cases related to `knn` search we'd capture (was looking at the usages but I'm not very familiar with all cases).

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchRequestParser.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/vectors/KnnSearchRequestParser.java:198]: We should not update this class anylonger. Its deprecated and folks shouldn't use the `_knn_search` API.

File: server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java:103]: I see that this is required for when we transform the knnquerybuilder to a knnQuery. However, I don't think this should be nullable. The default & the provided values will not change during the lifetime of the request and are all known when this is created. 

So, instead of allowing this to be null, can we default it to `DEFAULT_SIZE`?

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1590]: It seems like this isn't being used at all within the KnnSearchBuilder. However, we can do better. 

We should have a `KnnSearchBuilder.Builder` that `KnnSearchBuilder.fromXContent` returns. That can have setters for `k` that gets adjusted after the last field is read when parsing the `SearchSourceBuilder`. 

This way we have a `KnnSearchBuilder.Builder#build(int)` where `int` is the request size (the default or user provided). And `KnnSearchBuilder` will never have `null` `k` nor null `num_candidates`.

**Review by benwtrent — COMMENTED** [review]
Great stuff!

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1590]: I see what you mean - Nice suggestion! 
Currently, `org.elasticsearch.search.builder.SearchSourceBuilder#parseXContent` doesn't do much outside of the `while` loop other than iterating through all objects. I guess we should add the call to the builder within this parsing section though, as only then the `SearchSourceBuilder` would have been finalized parsing. 

Do you think that we should hide all other public test usages/constructors of `KnnSearchBuilder` behind the builder (except from the one accepting a `StreamInput`), or just isolate this very specific use-case?

File: docs/reference/rest-api/common-parms.asciidoc
Comment (pmpailis) [reply docs/reference/rest-api/common-parms.asciidoc]: Updated in 87292171a3b40ddf035fb3ac51fa2fddc8ba043b to ```Needs to be greater than `k` or `size` if `k` is omitted,
and cannot exceed 10,000. ```

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/150_knn_search_missing_params.yml
Comment (pmpailis) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/150_knn_search_missing_params.yml]: Fair enough, removed in a1237a577d9c7095c9074d7887d845ca82f075fe!

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchRequestParser.java
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/vectors/KnnSearchRequestParser.java]: Rolled back all changes to `KnnSearch`  in 1aa708b41edcc59ee63c4f10c61978c9b7147a75

File: server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java]: Yeap, makes sense! Updated in 4082cf718f4999f9c1ee98a8b2a4ff04da6da845

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/150_knn_search_missing_params.yml
Comment (pmpailis) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/150_knn_search_missing_params.yml]: Same as above - removed in https://github.com/elastic/elasticsearch/commit/a1237a577d9c7095c9074d7887d845ca82f075fe

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1590]: > Do you think that we should hide all other public test usages/constructors of KnnSearchBuilder behind the builder (except from the one accepting a StreamInput), or just isolate this very specific use-case?

I think the ctor should be package private & we use a builder object for all `final` object values.

Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1590]: >  And KnnSearchBuilder will never have null k nor null num_candidates.

So, since the responsibility to fill out missing values will be moved over to the `Builder` on the node that received the original request, we wouldn't need any transport versions or backwards compatibility as well, right? The values will have already been filled when sending over the objects, so the behavior right after `builder#build` will be pretty much the same.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1590]: If the KnnSearchBuilder doesn't have null values, it won't need bwc serialization as nothing changes there. But, the KnnVectorQueryBuilder probably still needs it as `num_candidates` can be null when executed as knn-as-a-query

Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1590]: Yeap, knn-query would still need it. But we can take it out of knn-search. Getting on it : )

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1590]: hey @pmpailis are you blocked by me here?

Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1590]: Nope @benwtrent ! Was just taking a look at `toXContent` to understand under which circumstances this is invoked, and more specifically in the context of `knn-search`. Will have an update for that shortly :)

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java:524]: ❤️

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java:116]: This can all still be `final` right?

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryVectorBuilderTestCase.java
Comment (benwtrent) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryVectorBuilderTestCase.java:73]:
```suggestion
    protected KnnSearchBuilder parseKnnSearchBuilder(XContentParser parser) throws IOException {
```

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryVectorBuilderTestCase.java
Comment (benwtrent) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryVectorBuilderTestCase.java:87]:
```suggestion
            this::parseKnnSearchBuilder
```
Calling it `parse` is tricky IMO. But, that is a minor thing.

**Review by benwtrent — COMMENTED** [review]
Just a couple of things. Looking REALLY nice :)

File: server/src/main/java/org/elasticsearch/TransportVersions.java
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/TransportVersions.java]: Updated `KNN_K_NUMCANDS_AS_OPTIONAL_PARAMS` TransportVersion in 390aff2521de4c95f1a2e7260e7ed5cac6e4f1fa

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java]: Yeap - restored them in 390aff2521de4c95f1a2e7260e7ed5cac6e4f1fa

File: server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java:185]: Good stuff, we cannot serialize to older nodes and we should just indicate that the value must be provided, by passing any talk of old nodes, etc.

**Review by benwtrent — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java:185]: Thanks @benwtrent ! Updated the exception message to be a bit more clear on why it's now failing.
```
"[" + NUM_CANDS_FIELD.getPreferredName() + "] field was a mandatory field in previous releases and is required to be non-null by some nodes in this cluster. Please make sure to provide the parameter as part of the request."
```

**Pull request merged**
